File: baekjoon/StringQuestion/047-17609-회문.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.BufferedWriter;
import java.io.OutputStreamWriter;

public class Main {
    public int palinedrome(String input) throws Exception {
        int lp = 0;
        int rp = input.length() - 1;
        int tlp, trp;
        Boolean isClean = true;
        while (lp < rp) {
            if (input.charAt(lp) != input.charAt(rp)) {
                if (input.charAt(lp+1) == input.charAt(rp)) {
                    tlp = lp + 1;
                    trp = rp;
                    while (tlp < trp) {
                        if (input.charAt(tlp) != input.charAt(trp)) {
                            isClean = false;
                            break;
                        }
                        tlp++;
                        trp--;
                    }
                    if (isClean) return 1;
                }

                isClean = true;
                if (input.charAt(lp) == input.charAt(rp-1)) {
                    tlp = lp;
                    trp = rp - 1;
                    while (tlp < trp) {
                        if (input.charAt(tlp) != input.charAt(trp)) {
                            isClean = false;
                            break;
                        }
                        tlp++;
                        trp--;
                    }
                    if (isClean) return 1;
                }
                
                return 2;
            }

            lp++;
            rp--;
        }

        return 0;
    }


    public void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringBuilder sb = new StringBuilder();

        int T = Integer.parseInt(br.readLine());
        for (int i = 0; i < T; i++) {
            sb.append(palinedrome(br.readLine())).append('\n');
        }

        bw.write(sb.toString());
        bw.flush();
    }


    public static void main(String[] args) throws Exception {
        new Main().solution();
    }
}

File: baekjoon/StringQuestion/047-17609-회문.py
import sys


def palindrome(string: str) -> int:
    lp = 0
    rp = len(string) - 1
    moveLp = True
    moveRp = True
    while lp < rp:
        if string[lp] != string[rp]:
            if string[lp+1] == string[rp]:
                tlp = lp + 1
                trp = rp
                while tlp < trp:
                    if string[tlp] != string[trp]:
                        moveLp = False
                        break
                    tlp += 1
                    trp -= 1
                if moveLp: return 1

            if string[lp] == string[rp-1]:
                trp = rp - 1
                tlp = lp
                while tlp < trp:
                    if string[tlp] != string[trp]:
                        moveRp = False
                        break
                    tlp += 1
                    trp -= 1
                if moveRp: return 1

            return 2
        
        lp += 1
        rp -= 1
    
    return 0


def main():
    input = sys.stdin.readline
    T = int(input())
    for _ in range(T):
        string = input().rstrip()
        print(palindrome(string))


if __name__ == '__main__':
    main()